Content update - Routed events overview (user story 1878475) (#1367)
dotnet/docs-desktop: dotnet-desktop-guide/net/wpf/events/routed-events-overview.md

### Top-level scenarios for routed events

Here are some of the scenarios that motivated the routed event concept, and distinguish it from a typical CLR event:

- **Control composition and encapsulation**: Various controls in WPF have a rich content model. For example, you can place an image inside a <xref:System.Windows.Controls.Button>, which effectively extends the visual tree of the button. But, the added image mustn't break the hit-test behavior of the button, which needs to respond when a user clicks the image pixels.

> - **Singular handler attachment points**: In Windows Forms, you have to attach the same handler multiple times to process events raised on multiple elements. Routed events enable you to attach that handler only once, as shown in the previous XAML example. Event handler logic can determine where the event came from. For example, this handler is applicable to the previously shown XAML element tree:

Reviewer: I don't much like this. Back when WPF came out, the only other managed UI stack was Windows Forms. You had a lot of people converting from it, so a statement like this made sense. It no longer makes sense to assume people know what WinForms is or how it even works. I tried to edit this to make the benefit clearer. How about this?

```suggestion
- **Singular handler attachment points**: You could register a handler with each button's `Click` event, but, routed events enables you to attach a single handler. This enables you to change element tree under the singular handler, such as adding or removing more buttons, without having to register each button's `Click` event. When the `Click` event is raised, handler logic can determine where the event came from. For example, this handler is applicable to the previously shown XAML element tree:
```

Author: Looks good, just added in a reference to the previous XAML example.

:::code language="xaml" source="./snippets/routed-events-overview/csharp/MainWindow.xaml" id="AddHandler_UnqualifiedEventName":::

If the listener is a different element than the element that raised the event, it's possible that the event isn't a member of the listener's class. In that case, you must use the qualified event name, in the form of `<owner type>.<event name>`. For example, because the <xref:System.Windows.Controls.StackPanel> class doesn't implement the <xref:System.Windows.Controls.Primitives.ButtonBase.Click> event, you'll need to use the qualified event name:

:::code language="xaml" source="./snippets/routed-events-overview/csharp/MainWindow.xaml" id="AddHandler_QualifiedEventName":::

The XAML syntax for adding routed event handlers is the same as for standard CLR event handlers. For more information about adding event handlers in XAML, see [XAML in WPF](../xaml/index.md).

> [!NOTE]

> > The element that originally raised a routed event is identified as the <xref:System.Windows.RoutedEventArgs.Source?displayProperty=nameWithType> in the event handler parameters. The event listener is the element where the event handler is attached and invoked, and is identified as the [sender](xref:System.Windows.RoutedEventHandler) in the event handler parameters.

Reviewer: Do you think we should delete this and add any missing info the `## Attach and implement a routed event handler` section?

Author: Good idea! Done.

To attach an event handler for a routed event to an element using code, you generally have two options:

- Directly call the <xref:System.Windows.UIElement.AddHandler%2A> method. Routed event handlers can always be attached this way. This example attaches a `Click` event handler to a button using the `AddHandler` method:

  :::code language="csharp" source="./snippets/routed-events-overview/csharp/MainWindow.xaml.cs" id="AddHandler_ToAttachHandler":::
  :::code language="vb" source="./snippets/routed-events-overview/vb/MainWindow.xaml.vb" id="AddHandler_ToAttachHandler":::

- If the routed event implements a CLR event wrapper, use language-specific event syntax to add event handlers just as you would for a standard CLR event. Most existing WPF routed events implement the CLR wrapper, thus enabling language-specific event syntax. This example attaches a `Click` event handler to a button using language specific syntax:

  :::code language="csharp" source="./snippets/routed-events-overview/csharp/MainWindow.xaml.cs" id="LanguageSpecificSyntax_ToAttachHandler":::

>   :::code language="vb" source="./snippets/routed-events-overview/vb/MainWindow.xaml.vb" id="LanguageSpecificSyntax_ToAttachHandler":::

Reviewer: I know this is a bit more involved than just scrubbing the article, but I feel like it may be more beneficial for our customers to see this implemented as the StackPanel handling the button click instead of a button handling the button click. The demonstration doesn't seem to show anything about the "routed" aspect.

Author: The existing AddHandler example mirrors the CLR event example. I've added the routed event scenario you mentioned (attach to StackPanel) as an additional example.

- Attach the handler in code-behind using the <xref:System.Windows.UIElement.AddHandler(System.Windows.RoutedEvent,System.Delegate,System.Boolean)> overload, with the `handledEventsToo` parameter set to `true`.

- Set the <xref:System.Windows.EventSetter.HandledEventsToo%2A> attribute in an `EventSetter` to `true`.

The concept of `Handled` might affect how you design your application and code your event handlers. You can conceptualize `Handled` as a simple protocol for processing of routed events. How you use this protocol is up to you, but the expected use of the `Handled` parameter is:

- If a routed event is marked as handled, then it doesn't need to be handled again by other elements along the route.

- If a routed event isn't marked as handled, then listeners earlier in the event route don't have a handler for the event, or none of the registered handlers responded to the event in a way that justifies marking the event as handled. Handlers on the current listener have three possible courses of action:

  - Take no action at all. The event remains unhandled and routes to the next listener in the tree.

>   - Execute code in response to the event, but not to an extent that justifies marking the event as handled. The event remains unhandled and routes to the next listener in the tree.
> 
>   - Execute code in response to the event, to an extent that justifies marking the event as handled. Mark the event as handled in the event data. The event still routes to the next listener in the tree, but most listeners won't invoke further handlers. The exception is listeners with handlers that were specifically registered with `handledEventsToo` set to `true`.

Reviewer: Should use Run instead of Execute. It's better for translation
```suggestion
  - Run code in response to the event, but not to an extent that justifies marking the event as handled. The event remains unhandled and routes to the next listener in the tree.

  - Run code in response to the event, to an extent that justifies marking the event as handled. Mark the event as handled in the event data. The event still routes to the next listener in the tree, but most listeners won't invoke further handlers. The exception is listeners with handlers that were specifically registered with `handledEventsToo` set to `true`.
```

> For more information about marking an event as handled and class handling of routed events, see [Marking routed events as handled, and class handling](marking-routed-events-as-handled-and-class-handling.md).

Reviewer: This could be simplified
```suggestion
For more information about handling routed events, see [Marking routed events as handled, and class handling](marking-routed-events-as-handled-and-class-handling.md).
```

A preview input event that's marked as handled won't invoke any normally registered event handlers for the remainder of the preview route, and the paired bubbling event won't be raised. This handling behavior is useful for composite controls designers who want hit-test based input events or focus-based input events to be reported at the top-level of their control. Top-level elements of the control have the opportunity to class-handle preview events from control subcomponents in order to "replace" them with a top-level control-specific event.

To illustrate how input event processing works, consider the following input event example. In the following tree illustration, `leaf element #2` is the source of both the `PreviewMouseDown` and `MouseDown` paired events:

:::image type="content" source="./media/routed-events-overview/input-event-routing.png" border="true" alt-text="Event routing diagram.":::

The order of event processing following a mouse-down action on leaf element #2 is:

> 1. `PreviewMouseDown` tunneling event on the root element.
> 1. `PreviewMouseDown` tunneling event on intermediate element #1.
> 1. `PreviewMouseDown` tunneling event on leaf element #2, which is the source element.
> 1. `MouseDown` bubbling event on leaf element #2, which is the source element.
> 1. `MouseDown` bubbling event on intermediate element #1.
> 1. `MouseDown` bubbling event on the root element.

Reviewer: Just a style nit for myself :)

```suggestion
01. `PreviewMouseDown` tunneling event on the root element.
01. `PreviewMouseDown` tunneling event on intermediate element #1.
01. `PreviewMouseDown` tunneling event on leaf element #2, which is the source element.
01. `MouseDown` bubbling event on leaf element #2, which is the source element.
01. `MouseDown` bubbling event on intermediate element #1.
01. `MouseDown` bubbling event on the root element.
```
